app/src/test/kotlin/compiler/intermediate_form/ControlFlowGraphBuilderTest.kt

```diff
@@ -0,0 +1,144 @@
+package compiler.intermediate_form
+import compiler.common.reference_collections.referenceHashMapOf
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+class ControlFlowGraphBuilderTest {
+    private val regRead = Register()
+    private val regWrite = Register()
+
+    private val innerNode = IntermediateFormTreeNode.RegisterRead(regRead)
+
+    private val entryNode = IntermediateFormTreeNode.RegisterWrite(regWrite, innerNode)
+    private val secondNode = IntermediateFormTreeNode.NoOp()
+    private val conditionalTrueNode = IntermediateFormTreeNode.NoOp()
+    private val conditionalFalseNode = IntermediateFormTreeNode.NoOp()
+
+    private val expectedCFG = ControlFlowGraph(
+        treeRoots = listOf(entryNode, secondNode, conditionalTrueNode, conditionalFalseNode),
+        entryTreeRoot = entryNode,
+        unconditionalLinks = referenceHashMapOf(entryNode to secondNode),
+        conditionalTrueLinks = referenceHashMapOf(secondNode to conditionalTrueNode),
+        conditionalFalseLinks = referenceHashMapOf(secondNode to conditionalFalseNode)
+    )
+
+    @Test
+    fun `test pass entryTreeRoot in constructor`() {
+        val cfg = ControlFlowGraphBuilder(entryNode).build()
+
+        assertEquals(cfg.entryTreeRoot, entryNode)
+        assertEquals(cfg.treeRoots, listOf(entryNode))
+        assertEquals(cfg.unconditionalLinks, referenceHashMapOf())
+        assertEquals(cfg.conditionalFalseLinks, referenceHashMapOf())
+        assertEquals(cfg.conditionalTrueLinks, referenceHashMapOf())
+    }
+
+    @Test
+    fun `test makeRoot`() {
+        val cfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.makeRoot(entryNode)
+
+        val cfg = cfgBuilder.build()
+        assertEquals(ControlFlowGraphBuilder(entryNode).build(), cfg)
+    }
+
+    @Test
+    fun `test makeRoot when already specified`() {
+        val cfgBuilder = ControlFlowGraphBuilder(entryNode)
+        assertFailsWith<IncorrectControlFlowGraphError> { cfgBuilder.makeRoot(entryNode) }
+    }
+
+    @Test
+    fun `test addLink from null sets entryTreeRoot`() {
```

Comment:
Name doesn't describe the test well. We test overall functionality of `addLink` here, not only its behavior when from = null.